### PR TITLE
Namespace flag allows tests with no ns creation

### LINKED ISSUE #6: Support --namespace flag
KUTTL currently always creates namespaces for each test step, but some users may want to run their tests in a specific namespace or target an existing deployment. This would also be useful for running `kudo test` in less privileged environments where it is unable to create namespaces.

It is possible today to target static namespaces by specifying them on any commands or manifests, but it would be useful to add a `--namespace` flag to kudo test that controls the namespace used.

For example, if `--namespace` is set on `kudo test`, instead of creating namespaces for each test step and using the created namespace as the default for each resource, `kudo test` will set the default namespace to the `--namespace` flag and use that instead.

### PR DESCRIPTION
**Pull request by kensipe**
`test --namespace` support provided
rules:
1. The default namespace is ""
2. If default namespace, then tests will create their own namespaces for each tests (previous behavior)
3. if the namespace is provided then no namespaces will be created or deleted and the tests will run in that namespace.

Simple test run with "temp" ns
```
go run cmd/kubectl-kuttl/main.go test --start-control-plane=false ~/projects/playground/e2e/ -n temp
=== RUN   kuttl
    kuttl: harness.go:357: starting setup
    kuttl: harness.go:228: running tests using configured kubeconfig.
    kuttl: harness.go:303: running tests
    kuttl: harness.go:69: going to run test suite with timeout of 120 seconds for each step
=== RUN   kuttl/harness
=== RUN   kuttl/harness/pod
    kuttl/harness/pod: logger.go:42: 13:27:56 | pod | Ignoring pod.yaml as it does not match file name regexp: ^(\d+)-([^.]+)(.yaml)?$
=== PAUSE kuttl/harness/pod
=== CONT  kuttl/harness/pod
    kuttl/harness/pod: logger.go:42: 13:27:56 | pod/0-install | starting test step 0-install
    kuttl/harness/pod: logger.go:42: 13:27:56 | pod/0-install | Pod:temp/static-web created
    kuttl/harness/pod: logger.go:42: 13:27:58 | pod/0-install | test step completed 0-install
    kuttl/harness/pod: logger.go:42: 13:27:58 | pod | pod events from ns temp:
    kuttl/harness/pod: logger.go:42: 13:27:58 | pod | 2020-06-16 13:27:56 -0500 CDT	Normal	Scheduled	Successfully assigned temp/static-web to kind-control-plane
    kuttl/harness/pod: logger.go:42: 13:27:58 | pod | 2020-06-16 13:27:56 -0500 CDT	Normal	Pulling	Pulling image "nginx"
    kuttl/harness/pod: logger.go:42: 13:27:58 | pod | 2020-06-16 13:27:57 -0500 CDT	Normal	Pulled	Successfully pulled image "nginx"
    kuttl/harness/pod: logger.go:42: 13:27:58 | pod | 2020-06-16 13:27:57 -0500 CDT	Normal	Created	Created container web
    kuttl/harness/pod: logger.go:42: 13:27:58 | pod | 2020-06-16 13:27:57 -0500 CDT	Normal	Started	Started container web
    kuttl: harness.go:342: run tests finished
--- PASS: kuttl (3.00s)
    --- PASS: kuttl/harness (0.00s)
        --- PASS: kuttl/harness/pod (2.38s)
PASS

```

Signed-off-by: Ken Sipe <kensipe@gmail.com>
Fixes #6

### CODE REVIEW

File: pkg/apis/testharness/v1beta1/test_types.go [hunk 1]
```diff
@@ -55,6 +55,10 @@ type TestSuite struct {
 
 	// ReportFormat determines test report format (JSON|XML|nil) nil == no report
 	ReportFormat *report.Type
+	// Namespace defines the namespace to use for tests
+	// The value "" means to auto-generate tests namespaces, these namespaces will be created and removed for each test
+	// Any other value is the name of the namespace to use and it must be created prior to tests running
```
Comment (porridge) [inline pkg/apis/testharness/v1beta1/test_types.go:60]:
```suggestion
	// Any other value is the name of the namespace to use and it must exist prior to running tests
```

File: pkg/kuttlctl/cmd/test.go [hunk 2]
```diff
@@ -161,6 +162,10 @@ For more detailed documentation, visit: https://kudo.dev/docs/testing`,
 				options.ArtifactsDir = artifactsDir
 			}
 
+			if isSet(flags, "namespace") {
+				options.Namespace = namespace
```
Comment (porridge) [inline pkg/kuttlctl/cmd/test.go:166]: An interesting corner case is what happens when user specifies `--namespace ''` (this can happen intentionally or due to a misspelled variable name in a script for example).
Perhaps we should validate that the flag value is non-empty if specified explicitly?

File: pkg/test/case.go [hunk 3]
```diff
@@ -42,6 +43,11 @@ type Case struct {
 
 // DeleteNamespace deletes a namespace in Kubernetes after we are done using it.
 func (t *Case) DeleteNamespace(namespace string) error {
+	// if system defined namespace is "" we create and delete, otherwise don't
+	if t.Namespace != "" {
+		return nil
```
Comment (porridge) [inline pkg/test/case.go:48]:
```suggestion
// DeleteNamespace deletes the auto-created namespace in Kubernetes after we are done using it.
func (t *Case) DeleteNamespace(namespace string) error {
	if t.Namespace != "" {
		t.Loggger.Log("Skipping deletion of user-supplied namespace:", namespace)
		return nil
```

File: pkg/test/case.go [hunk 4]
```diff
@@ -42,6 +43,11 @@ type Case struct {
 
 // DeleteNamespace deletes a namespace in Kubernetes after we are done using it.
 func (t *Case) DeleteNamespace(namespace string) error {
```
Comment (porridge) [inline pkg/test/case.go:45]: It is really confusing to a casual reader right now is that we have the `namespace` arg and `t.Namespace`. They match usually, but not when `t.Namespace` has the sentinel value. We document the design in comments, which is fragile.

I think we should aim for self-documenting code whose semantics is obvious. How about making `Namespace` a `struct` with `Name string` and `Autocreated bool`. In that case we can get rid of the `namespace` arg altogether.

Comment (porridge) [reply pkg/test/case.go:45]: Same for `CreateNamespace`

File: pkg/test/case.go (hunk 3 above)
Comment (porridge) [reply pkg/test/case.go:48]: Same for `CreateNamespace`

Comment (kensipe) [reply pkg/test/case.go:48]: yeah... I like it!

File: pkg/test/case.go (hunk 4 above)
Comment (kensipe) [reply pkg/test/case.go:45]: I think I like the struct suggestion.. but requires some thinking... My initial thought is I'm expecting some refactoring shortly and perhaps it should go into that work.  If we don't do it here for this PR... I will make sure to capture it in an issue (so it is not lost).   also...  as discussed in a side conversation... we decided to create the supplied ns if it doesn't exist, on the premise that if the ns doesn't exist, the test would fail.  The main reason to not create a ns, is the user running kuttl may not have the rights to create a ns.  If the tests were going to fail anyway... attempting to create the ns will fail for another reason but if they have rights has the chance of succeeding.   This would change the nature of these checks... at min everything is potentially "auto-created" reducing the value of the suggested struct.  However perhaps there is room for something similar...  I'm with you!  self-documenting code is better... also localizing decisions is better.

File: pkg/kuttlctl/cmd/test.go (hunk 2 above)
Comment (kensipe) [reply pkg/kuttlctl/cmd/test.go:166]: nice... agree!   --namespace="" explicit should mean cluster scope which we do not allow and should explicit check for... thanks @porridge

File: pkg/test/case.go (hunk 4 above)
Comment (kensipe) [reply pkg/test/case.go]: @porridge turns out... this looks really good with the stucts.  thank you sir!

File: pkg/test/case.go [hunk 5]
```diff
@@ -33,16 +34,27 @@ type Case struct {
...
 
 	Client          func(forceNew bool) (client.Client, error)
 	DiscoveryClient func() (discovery.DiscoveryInterface, error)
 
 	Logger testutils.Logger
 }
 
+type namespace struct {
+	Name        string
+	AutoCreated bool
```
Comment (porridge) [inline pkg/test/case.go:47]: The code is clearer now, I am still concerned that `Case struct` contains a `Namespace` field, and there is also an `ns` variable. The distinction between their purpose is not clear at first sight.

If you don't think we could eliminate one of these, maybe we could at least rename `Case.Namespace` to something line `Case.NamespaceSpec` to make it clearer that it is a recipe for discovering/generating the actual namespace to use?

File: pkg/test/case.go [hunk 6]
```diff
@@ -165,7 +200,26 @@ func (t *Case) Run(test *testing.T, tc *report.Testcase) {
...
-	t.CollectEvents(ns)
+	t.CollectEvents(ns.Name)
+}
+
+func (t *Case) determineNamespace() (*namespace, error) {
+	ns := &namespace{
+		Name:        t.Namespace,
+		AutoCreated: false,
+	}
+	//ns := t.Namespace
```
Comment (porridge) [inline pkg/test/case.go:211]: Is this line a genuine comment or a remnant of debugging? If the former than I have to admit it's not very clear.

File: pkg/test/case.go (hunk 5 above)
Comment (kensipe) [reply pkg/test/case.go:47]: it isn't a namespace "spec"... but I get your point... the issue to be considered is if we make `case.namespace` a `namespace` type, than do we make `TestSuite` of that type... I'm real hesitate at this point, because this is externally exposed in the test suite file.  Let me think on this one.

Comment (kensipe) [reply pkg/test/case.go:47]: hmm... after more thought I think I understand why your suggestion better on naming... I'm real hesitate to use the "spec" word as it carries different meaning in the k8s world.  still thinking...
